### PR TITLE
Add JSON hints list and exhaustive ignore list generation

### PR DESCRIPTION
**Pull request by tbagrel1**
With (merged) PR [#1056](https://github.com/ndmitchell/hlint/pull/1056), we now have a way to generate a Markdown list of all built-in hints (at least, the ones present in tests) and current custom ones.

This PR aims to bring two new commands `--generate-summary-json` and `--generate-ignore-all`.

+ `--generate-summary-json` creates an equivalent of the markdown summary of `--generate-summary`, but in JSON format (so that the list can be exported in various formats, and be used programatically). ~Only built-in hints are reported with this command. ~ All hints are now reported
+ ~`--generate-ignore-all` creates a hlint config file (in the `.hlint.yaml` format) in which all built-in hints are disabled by default (and then the user can opt-in by replacing `ignore` by `warn`)~
+ `--generate-exhaustive-config[=LEVEL]` creates a hlint config file (in the `.hlint.yaml` format) in which **all** hints are set to the specified severity level

All examples related to each hint are ~now collected in both markdown and JSON summary~ collected in the JSON summary, but the Markdown summary only shows one example per hint.

I also added the generated JSON and .yaml files in `data` (I might be easier for the user to just copy them instead of cloning `hlint` + generating them).

**Possible improvements:**

+ Add examples/description above each hint in the "ignore all" file
+ **DONE** ~Add a `--generate-warn-all` command (in which all hints are set to `warn` by default), or replace both commands with `--generate-exhaustive-hlintyaml [LEVEL (default: warn)]`~

I'm strongly in favor of having a way to generate an exhaustive config file, so that users can configure hlint once for all. Without that option, they need to edit their hlint config every time they encounter a hint that they dislike.

**Current issues:**

+ A few hint names appears twice in the summary/ignore list. I think that it is caused by two hints sharing the same name, but having different {severity/isRefactoring/category}. **Partially solved:** hint names are now sorted and deduped in the generated config files

Feel free to edit this PR directly on the branch from my fork.

### CODE REVIEW

**Comment by tbagrel1** [comment]
> I think only showing one example is quite helpful for the markdown, otherwise it gets much harder to read the markdown as it gets a fair chunk longer. Other than that, this seems like a good addition - thanks!

I reverted this change, and added some improvements along the way. Let me know if I need to adjust something :)

**Comment by tbagrel1** [comment]
Just to be sure, can we actually set the severity level of LHS/RHS hints in the `.hlint.yaml` config file? At first, I thought that LHS/RHS hints were user-defined ones (but it seems that a lot of them are shipped with Hlint).

**Review by ndmitchell — COMMENTED** [review]
Thanks for the diff, and sorry it's taken so long to review. Looks fine, just a few questions

For the hints.md, it's really hard to review, and I would like to understand what has happened to the file. Is it pretty much just a reordering? Or are there any other details changed?

What would be the use case for data/error-all? Or suggest-all? I can see the use of ignore-all, but not the others. Even if we provide the feature, is it worth bothering to include the others in the repo?

**Comment by ndmitchell** [comment]
> Just to be sure, can we actually set the severity level of LHS/RHS hints in the .hlint.yaml config file? 

Yes, and many of the LHS/RHS hints are shipped by HLint - as many as we can

File: src/CmdLine.hs [hunk 1]
```diff
@@ -125,7 +126,9 @@ data Cmd
         ,cmdRefactorOptions :: String   -- ^ Options to pass to the `refactor` executable.
         ,cmdWithRefactor :: FilePath    -- ^ Path to refactor tool
         ,cmdIgnoreGlob :: [FilePattern]
-        ,cmdGenerateSummary :: [FilePath]  -- ^ Generate a summary of available hints
+        ,cmdGenerateMdSummary :: [FilePath]  -- ^ Generate a summary of available hints, in Markdown format
+        ,cmdGenerateJsonSummary :: [FilePath]  -- ^ Generate a summary of built-in hints, in JSON format
+        ,cmdGenerateExhaustiveConf :: [Severity]  -- ^ Generate a hlint config file with all built-in hints set to the specified level
```
[the remark this answers is not in the document]
Comment (tbagrel1) [reply src/CmdLine.hs:131]: I'm not sure to understand. I tried to copy how `generateSummary` worked for `generateJsonSummary` and `generateExhaustiveConf`. And then, I renamed every `generateSummary` occurence to `generateMdSummary`. So I expect it to work the same way as before (but I must admit that I had some trouble understanding the CLI-related structure).

**Comment by tbagrel1** [comment]
> Thanks for the diff, and sorry it's taken so long to review. Looks fine, just a few questions
> 
> For the hints.md, it's really hard to review, and I would like to understand what has happened to the file. Is it pretty much just a reordering? Or are there any other details changed?

It should be just a reordering. Also, the displayed example might have changed (because example ordering for each rule might have changed).

> What would be the use case for data/error-all? Or suggest-all? I can see the use of ignore-all, but not the others. Even if we provide the feature, is it worth bothering to include the others in the repo?

It would be nice to have warn-all and ignore-all in the repo (because they correspond to explicit opt-in/opt-out for all rules). But I agree, error-all and suggest-all are not really needed.

___

Do you want me to adress these changes, or would you like to adjust yourself the final details at your convenience?

**Comment by aspiwack** [comment]
@ndmitchell what can @tbagrel1 do to work toward merging this PR?

**Comment by ndmitchell** [comment]
Sorry for the delay. Personal reasons mean I'll probably merge this in one week and make a fairly immediate HLint release thereafter. No action required on anyone else's behalf.

**Comment by ndmitchell** [comment]
Currently there is a section in the HLint README that says for developers:

> After changing hints, you will need to regenerate the [hints.md](hints.md) file with `hlint --generate-summary`.

In addition there is a test at https://github.com/ndmitchell/hlint/blob/cd529ad925bd40579bd0bef06f86cfae6c0da783/travis.hs#L8 that the generated summary is up to date. Alas, regenerating that summary is a real pain. I guess the instructions and tests need to change to take into account the new commands? To what?

**Comment by tbagrel1** [comment]
I just added a commit that should update the README and travis tests to take new commands into account. Let me know if this sounds good to you.

Regeneration of `{error,warn,ignore,suggest}-all.yaml` files is not pretty, because the associated command takes a severity level instead of a path as its argument, so we can't put them directly in the `data/` directory.

> Alas, regenerating that summary is a real pain.

I didn't understand this part. Is something required on my end?

**Comment by ndmitchell** [comment]
My concern is that currently we regenerate the `hints.md` file summary, and that's a pain. It leads to conflicts (as has happened on this diff), it requires an extra step making drive-by contributions to `hlint.yaml` harder, it requires people to submit diffs against HEAD hlint rather than a released version. But it has benefits because having a list of hints available to users makes it easy to see what you are getting and it's easily viewable on the web without us building some server-side stuff. It's a trade off, but seems generally positive.

This diff introduces a mechanism for generating summaries (fine), plus 5 extra files. You've marked that `# optional: regenerate auxiliary hints files in data/` - but I'm not sure it is optional - if we're providing files, they should be correct. It's also quite a bit more effort to generate these additional files, costs borne by every developer. So I'm wondering if those costs are worth it, or simply telling users to generate them if they need them is a better trade-off?

**Comment by tbagrel1** [comment]
Oh I see, thanks!
I marked this step as "optional" because I thought it was intended for end users who might want to add hints. I didn't consider contributors; but I agree, this step is no longer optional for contributors.
I also agree with the fact that it is a bit painful to renegerate this files manually for each contribution.
I think you are better informed than I am about the cost vs benefit of having the `hints.json` and `*-all.yaml` files in the repository. You can freely remove them from the PR if you think that their inclusion is not worth it :)

**Comment by ndmitchell** [comment]
I think the best thing is to not have these files generated and checked in for now. Can you please remove them, regenerate hlint.md, then I'll merge. You'll need to remove the README.md notes and travis.hs too now they aren't on by default.

File: src/CmdLine.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tbagrel1) [reply src/CmdLine.hs:131]: Now I understand the question :)

The output for cmdGenerateExhaustiveConf goes to `${severity}-all.yaml`, in the current directory.

**Comment by tbagrel1** [comment]
I hope that it looks good to you now :slightly_smiling_face:

**Comment by tbagrel1** [comment]
Do you know whether this PR can be merged, or if there is anything left to do?

**Comment by ndmitchell** [comment]
Sorry this took so long. The main cause was my unavailability due to personal circumstances and illness. I'm going to try address that with #1395

**Comment by ndmitchell** [comment]
I renamed `--generate-exhaustive-config` as to `--generate-config` since it's only exhaustive for now, it might stop being exhaustive in the near future. It's also easier to spell.

**Comment by ndmitchell** [comment]
The custom instance for `instance Show Severity` broke all the tests, so I have reverted that piece.

**Pull request merged**